### PR TITLE
Merge fuse_conv_and_bn and fuse_deconv_and_bn, and provide type check…

### PR DESCRIPTION
**Pull request by ghost**
Thank you 🙏 for your contribution to [Ultralytics](https://ultralytics.com) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. **Check for Existing Contributions**: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.

2. **Link Related Issues**: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.

3. **Elaborate Your Changes**: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.

4. **Ultralytics Contributor License Agreement (CLA)**: To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    _I have read the CLA Document and I hereby sign the CLA_

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing). Your adherence to these guidelines ensures a faster and more effective review process.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced layer fusion functionality for n-dimensional convolutions and transposed convolutions with batch normalization.

### 📊 Key Changes
- Introduced the `test_fuse.py` file containing tests for fusing convolution and batch normalization layers.
- Amended the `nn/tasks.py` file to employ the updated fusion function `fuse_conv_and_bn`.
- Updated `torch_utils.py` with a more generalized `fuse_nd_conv_and_bn` function to handle n-dimensional convolutions.
- Removed the redundant `fuse_deconv_and_bn` function.

### 🎯 Purpose & Impact
- 🛠 **Improves test coverage**: New tests ensure that the fusion of convolutional layers with batch normalization operates correctly.
- 🔧 **Streamlines code**: Unifies the functions for fusing convolutional layers, whether they are transposed or not, simplifying the codebase.
- 🌐 **Broader applicability**: The new function can fuse n-dimensional convolution layers, increasing its usefulness for various neural network architectures.
- ⚡ **Enhances performance**: Fusing layers can lead to performance gains by reducing computational overhead and memory usage during inference.

### CODE REVIEW

**Comment by YMCAlan** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by YMCAlan** [comment]
This is my test code. If there are any questions or bugs that need to be addressed, please let me know.
```
import unittest
import torch
import torch.nn as nn
from torchfuse.functional import fuse_conv_and_bn


class TestConvBNFusion(unittest.TestCase):
    def setUp(self):
        self.conv1d = nn.Conv1d(3, 64, kernel_size=3, padding=1).eval()
        self.bn1d = nn.BatchNorm1d(64).eval()

        self.conv2d = nn.Conv2d(3, 64, kernel_size=3, padding=1).eval()
        self.bn2d = nn.BatchNorm2d(64).eval()

        self.conv3d = nn.Conv3d(3, 64, kernel_size=3).eval()
        self.bn3d = nn.BatchNorm3d(64).eval()

        self.trans_conv1d = nn.ConvTranspose1d(64, 3, kernel_size=3, padding=1).eval()
        self.trans_bn1d = nn.BatchNorm1d(3).eval()

        self.trans_conv2d = nn.ConvTranspose2d(64, 3, kernel_size=3, padding=1).eval()
        self.trans_bn2d = nn.BatchNorm2d(3).eval()

        self.trans_conv3d = nn.ConvTranspose3d(64, 3, kernel_size=3).eval()
        self.trans_bn3d = nn.BatchNorm3d(3).eval()

    def test_fuse_conv_bn(self):
        # Test the function for 1D Convolution and BatchNormalization
        fused_conv_1d = fuse_conv_and_bn(self.conv1d, self.bn1d)
        self.assertIsInstance(fused_conv_1d, nn.Conv1d)

        # Test the function for 2D Convolution and BatchNormalization
        fused_conv_2d = fuse_conv_and_bn(self.conv2d, self.bn2d)
        self.assertIsInstance(fused_conv_2d, nn.Conv2d)

        # Test the function for 3D Convolution and BatchNormalization
        fused_conv_3d = fuse_conv_and_bn(self.conv3d, self.bn3d)
        self.assertIsInstance(fused_conv_3d, nn.Conv3d)

        # Test the function for 1D Transposed Convolution and BatchNormalization
        fused_trans_conv_1d = fuse_conv_and_bn(self.trans_conv1d, self.trans_bn1d)
        self.assertIsInstance(fused_trans_conv_1d, nn.ConvTranspose1d)

        # Test the function for 2D Transposed Convolution and BatchNormalization
        fused_trans_conv_2d = fuse_conv_and_bn(self.trans_conv2d, self.trans_bn2d)
        self.assertIsInstance(fused_trans_conv_2d, nn.ConvTranspose2d)

        # Test the function for 3D Transposed Convolution and BatchNormalization
        fused_trans_conv_3d = fuse_conv_and_bn(self.trans_conv3d, self.trans_bn3d)
        self.assertIsInstance(fused_trans_conv_3d, nn.ConvTranspose3d)

    def test_fuse_conv_bn_error(self):
        # Test error cases for mismatched types
        mismatched_combinations = [
            (self.conv1d, self.bn2d),
            (self.conv1d, self.bn3d),
            (self.conv2d, self.bn1d),
            (self.conv2d, self.bn3d),
            (self.conv3d, self.bn1d),
            (self.conv3d, self.bn2d),
            (self.trans_conv1d, self.trans_bn2d),
            (self.trans_conv1d, self.trans_bn3d),
            (self.trans_conv2d, self.trans_bn1d),
            (self.trans_conv2d, self.trans_bn3d),
            (self.trans_conv3d, self.trans_bn1d),
            (self.trans_conv3d, self.trans_bn2d),
        ]

        for conv, bn in mismatched_combinations:
            with self.subTest(conv=conv, bn=bn):
                with self.assertRaises(ValueError):
                    # This should raise an error due to mismatched types
                    fuse_conv_and_bn(conv, bn)

    def test_fuse_nd_conv_and_nd_bn(self):
        # Test the fused convolution for 1D Convolution and BatchNormalization
        fused_conv_1d = fuse_conv_and_bn(self.conv1d, self.bn1d)

        # Input a random batch
        x = torch.randn(1, 3, 6)  # (batch_size, in_channels, input_size)

        bn_output = self.bn1d(self.conv1d(x))
        fused_output = fused_conv_1d(x)

        # Ensure the fused_conv output matches conv and bn output
        error = (bn_output - fused_output).abs().mean()
        print(f"Error  (1D Conv): {error}")
        self.assertTrue(error < 1e-5)

        # Test the fused convolution for 2D Convolution and BatchNormalization
        fused_conv_2d = fuse_conv_and_bn(self.conv2d, self.bn2d)

        # Input a random batch
        x = torch.randn(1, 3, 6, 6)  # (batch_size, in_channels, height, width)

        bn_output = self.bn2d(self.conv2d(x))
        fused_output = fused_conv_2d(x)

        # Ensure the fused_conv output matches conv and bn output
        error = (bn_output - fused_output).abs().mean()
        print(f"Error  (2D Conv): {error}")
        self.assertTrue(error < 1e-5)

        # Test the fused convolution for 3D Convolution and BatchNormalization
        fused_conv_3d = fuse_conv_and_bn(self.conv3d, self.bn3d)

        # Input a random batch
        x = torch.randn(1, 3, 6, 6, 6)  # (batch_size, in_channels, depth, height, width)

        bn_output = self.bn3d(self.conv3d(x))
        fused_output = fused_conv_3d(x)

        # Ensure the fused_conv output matches conv and bn output
        error = (bn_output - fused_output).abs().mean()
        print(f"Error  (3D Conv): {error}")
        self.assertTrue(error < 1e-5)

    def test_fuse_nd_deconv_and_nd_bn(self):
        fused_trans_conv_1d = fuse_conv_and_bn(self.trans_conv1d, self.trans_bn1d)

        # Input a random batch
        x = torch.randn(1, 64, 6)  # (batch_size, in_channels, input_size)

        bn_output = self.trans_bn1d(self.trans_conv1d(x))
        fused_output = fused_trans_conv_1d(x)

        # Ensure the fused_conv output matches conv and bn output
        error = (bn_output - fused_output).abs().mean()
        print(f"Error (1D Transpose): {error}")
        self.assertTrue(error < 1e-5)

        # Test the fused convolution for 2D Transposed Convolution and BatchNormalization
        fused_trans_conv_2d = fuse_conv_and_bn(self.trans_conv2d, self.trans_bn2d)

        # Input a random batch
        x = torch.randn(1, 64, 6, 6)  # (batch_size, in_channels, height, width)

        bn_output = self.trans_bn2d(self.trans_conv2d(x))
        fused_output = fused_trans_conv_2d(x)

        # Ensure the fused_conv output matches conv and bn output
        error = (bn_output - fused_output).abs().mean()
        print(f"Error (2D Transpose): {error}")
        self.assertTrue(error < 1e-5)

        # Test the fused convolution for 3D Transposed Convolution and BatchNormalization
        fused_trans_conv_3d = fuse_conv_and_bn(self.trans_conv3d, self.trans_bn3d)

        # Input a random batch
        x = torch.randn(1, 64, 6, 6, 6)  # (batch_size, in_channels, depth, height, width)

        bn_output = self.trans_bn3d(self.trans_conv3d(x))
        fused_output = fused_trans_conv_3d(x)

        # Ensure the fused_conv output matches conv and bn output
        error = (bn_output - fused_output).abs().mean()
        print(f"Error (3D Transpose): {error}")
        self.assertTrue(error < 1e-5)


if __name__ == '__main__':
    unittest.main()

```

image

**Comment by YMCAlan** [comment]
recheck

**Comment by YMCAlan** [comment]
recheck

**Comment by YMCAlan** [comment]
Thank you for your response. I have already uploaded the test code and made some fixes to the previous code.

**Comment by glenn-jocher** [comment]
@YMCAlan thank you for updating your contribution with the test code and fixes. We'll review the changes in your Pull Request. Your efforts to improve the codebase are much appreciated! 🌟 If there are any further adjustments needed or if we have additional feedback, we'll communicate through the PR comments. Keep up the great work! 🚀
